clients/php/src/ZawgyiUnicodeMarkovModel.php

```diff
@@ -0,0 +1,231 @@
+        }
+
+        return 0;
+    }
+
+    /**
+     * The number of states in the Markov chain.
+     */
+    public static function getSize()
+    {
+        return self::NUM_STATES;
+    }
+
+    /**
+     * @param string $input
+     * @param bool $verbose
+     * @return float
+     */
+    public function predict(string $input, bool $verbose = false): float
+    {
+        if ($verbose) {
+            printf("Running detector on string: %s\n", $input);
+        }
+
+        // Start at the base state
+        $prevCp    = 0;
+        $prevState = 0;
+
+        $totalDelta     = 0.0;
+        $seenTransition = false;
+
+        for ($offset = 0; $offset <= mb_strlen($input);) {
+            if ($offset == mb_strlen($input)) {
+                $cp        = 0;
+                $currState = 0;
+            } else {
+                $character = mb_substr($input, $offset, 1);
+                $cp = self::ord_utf8($character);
+                $currState = self::getIndexForCodePoint($cp);
+            }
+            // Ignore 0-to-0 transitions
+            if ($prevState != 0 || $currState != 0) {
+                $delta = $this->classifier->getLogProbabilityDifference($prevState, $currState);
+                if ($verbose) {
+                    printf("U+%04X -> U+%04X: delta=%8.4f ", $prevCp, $cp, $delta);
+                    for ($i = 1; $i < abs($delta); $i++) {
+                        printf("!");
+                    }
+                    print("\n");
+                }
+                $totalDelta     += $delta;
+                $seenTransition = true;
+            }
+
+            $offset    += self::charCount($cp);
+            $prevCp    = $cp;
+            $prevState = $currState;
+        }
+
+        if ($verbose) {
+            printf("Final: delta=%.4f\n", $totalDelta);
+        }
+
+        // Special case: if there is no signal, return -Infinity,
+        // which will get interpreted by users as strong Unicode.
+        // This happens when the input string contains no Myanmar-range code points.
+        if (!$seenTransition)
+            return INF;
+
+        // result = Pz/(Pu+Pz)
+        //        = exp(logPz)/(exp(logPu)+exp(logPz))
+        //        = 1/(1+exp(logPu-logPz))
+        return 1.0 / (1.0 + exp($totalDelta));
+    }
+
+    public static function ord_utf8($c)
+    {
+        if (empty($c))
+            return;
+
+        $b0 = ord($c[0]);
+        if ($b0 < 0x80) {
+            return $b0;
+        }
+        $b1 = ord($c[1]);
+        if ($b0 < 0xE0) {
+            return (($b0 & 0x1F) << 6) + ($b1 & 0x3F);
+        }
+
+        return (($b0 & 0x0F) << 12) + (($b1 & 0x3F) << 6) + (ord($c[2]) & 0x3F);
+    }
+
+    # Determine the number of char values needed to represent the specified character (Unicode code point)
+    # if the code point is equal to or greater than 0x10000, then the method returns 2.
+    # otherwise, the method returns 1.
+    public static function charCount($codePoint)
```

Comment:
This is for UTF-16 handling in Java and JavaScript.  It is irrelevant in PHP, which is UTF-8.  Please delete this function.